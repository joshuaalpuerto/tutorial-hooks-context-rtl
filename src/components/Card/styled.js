import styled from 'styled-components'

export const DeleteButton = styled.span`
  &:after {
    color: #ccc;
    content: '\00d7';
    font-family: Arial, sans-serif;
    font-weight: 500;
    font-size: 24px;
    position: absolute;
    right: 10px;
    top: 5px;
    z-index: 1;
  }
`

export const IdWrapper = styled.div`
  background-color: ${({ bgColor }) => bgColor};
  color: #fff;
  padding: 5px;
`

export const ContentWrapper = styled.div`
  padding: 10px;
  padding-right: 80px;
  display: flex;
  flex-direction: column;

  & span {
    color: #ccc;
  }
`

export const Remove = styled.div`
  cursor: pointer;
  display: none;
`

export const Wrapper = styled.div`
  background-color: #fff;
  border: 2px solid #ccc;
  display: inline-flex;
  margin: 10px;
  padding: 0;
  position: relative;

  &:hover {
    ${/* sc-custom 'div' */ IdWrapper} {
      background-color: #1b998e;
    }
    ${/* sc-custom 'span' */ Remove} {
      display: block;
    }
  }

  @media only screen and (max-width: 768px) {
    display: flex;
    margin: 0;

    ${/* sc-custom 'span' */ Remove} {
      display: block;
    }
  }
`
