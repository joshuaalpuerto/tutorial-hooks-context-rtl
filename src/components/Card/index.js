import React from 'react'
import PropTypes from 'prop-types'
import { IdWrapper, ContentWrapper, Remove, Wrapper } from './styled'

export const DEFAULT_COLOR = '#80878B'
/**
 * Make renderDelete available for them to customize or not to provide just for display
 * @param {*} param0
 */
function Card({ id, color, heading, subheading, renderDelete }) {
  const bgColor = color || DEFAULT_COLOR
  return (
    <Wrapper data-testid="wrapper">
      <IdWrapper data-testid="id" bgColor={bgColor}>{id}</IdWrapper>
      <ContentWrapper>
        {heading}
        {subheading && <span data-testid="subheading"> {subheading} </span>}
      </ContentWrapper>
      <Remove data-testid="remove">{renderDelete}</Remove>
    </Wrapper>
  )
}

Card.defaultProps = {
  color: DEFAULT_COLOR,
  // renderDelete: null
}

Card.propTypes = {
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  heading: PropTypes.string.isRequired,
  renderDelete: PropTypes.node,
  subheading: PropTypes.string,
  color: PropTypes.string
}

export default Card
