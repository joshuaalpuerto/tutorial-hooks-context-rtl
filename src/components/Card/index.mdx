---
name: Card
---

import { Playground, Props } from 'docz'
import Card from './'
import { DeleteButton } from './styled'

# Card

<Props of={Card}/>

## Basic Usage

<Playground>
  <h2> Basic </h2>
  <Card
    id="1"
    heading="React"
    subheading="1 year"
  />
  <Card
    id="2"
    heading="NodeJS"
    subheading="1 year"
  />
  <br />
  <h2> With delete button </h2>
  <Card
    id="1"
    heading="React"
    subheading="1 year"
    renderDelete={() => <DeleteButton />}
  />
</Playground>
