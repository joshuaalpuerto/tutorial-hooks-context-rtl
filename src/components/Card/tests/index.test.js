import React from 'react'
import { render, fireEvent } from 'react-testing-library'

import Card, { DEFAULT_COLOR } from '../index'

import 'jest-styled-components'

describe('IdWrapper', () => {
  const props = {
    id: 1,
    heading: 'React',
    subheading: '1 year'
  }

  it('Should have default bgColor even null values submitted', () => {
    const { container, getByTestId } = render(<Card {...{ ...props, color: null }} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(getByTestId('id')).toHaveStyleRule('background-color', DEFAULT_COLOR)
  })

  it('should be background default #80878B', () => {
    const { container, getByTestId } = render(<Card {...{ ...props }} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(getByTestId('id')).toHaveStyleRule('background-color', DEFAULT_COLOR)
  })

  it('should have different background color if passes as prop', () => {
    const subProps = {
      ...props,
      color: 'lightgreen'
    }
    const { container, getByTestId } = render(<Card {...{ ...subProps }} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(getByTestId('id')).toHaveStyleRule('background-color', subProps.color)
  })
})

describe('ContentWrapper', () => {
  const props = {
    id: 1,
    heading: 'React',
    subheading: '1 year'
  }

  it('should have heading and subheading content', () => {
    const { getByText } = render(<Card {...{ ...props }} />)
    expect(getByText(props.heading)).toBeInTheDocument()
    expect(getByText(props.subheading)).toBeInTheDocument()
  })
})

describe('Remove', () => {
  const props = {
    id: 1,
    heading: 'React',
    subheading: '1 year'
  }

  it('should have "X" icon for deletion', () => {
    const subProps = {
      ...props,
      renderDelete: <div> X </div>
    }

    const { getByText } = render(<Card {...{ ...subProps }} />)
    // should be in the document
    expect(getByText('X')).toBeInTheDocument()
    // but it should be hidden
    expect(getByText('X')).not.toBeVisible()
  })

  /**
   * Skipped since it's too much unit testing as stated here
   * https://spectrum.chat/testing-library/general/how-do-i-test-styles-for-hover-events~108403f4-915b-4243-974c-c41af826b91d
   */
  it.skip('should show "X" on hover', () => {
    const subProps = {
      ...props,
      renderDelete: <div> X </div>
    }

    const { getByTestId } = render(<Card {...{ ...subProps }} />)
    const wrapper = getByTestId('wrapper')

    fireEvent.mouseOver(wrapper)

    // should be in the document
    expect(wrapper.querySelector('[data-testid="remove"]')).toHaveStyleRule('display', 'block')
  })

  it('Remove should be empty if renderDelete is not provided', () => {
    const { getByTestId } = render(<Card {...{ ...props }} />)
    expect(getByTestId('remove')).toBeEmpty()
  })
})

