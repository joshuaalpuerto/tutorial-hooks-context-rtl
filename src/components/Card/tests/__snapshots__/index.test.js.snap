// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdWrapper Should have default bgColor even null values submitted 1`] = `
.c2 {
  background-color: #80878B;
  color: #fff;
  padding: 5px;
}

.c3 {
  padding: 10px;
  padding-right: 80px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 span {
  color: #ccc;
}

.c5 {
  cursor: pointer;
  display: none;
}

.c0 {
  background-color: #fff;
  border: 2px solid #ccc;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 10px;
  padding: 0;
  position: relative;
}

.c0:hover .c1 {
  background-color: #1b998e;
}

.c0:hover .c4 {
  display: block;
}

@media only screen and (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    margin: 0;
  }

  .c0 .c4 {
    display: block;
  }
}

<div
  class="c0"
  data-testid="wrapper"
>
  <div
    class="c1 c2"
    data-testid="id"
  >
    1
  </div>
  <div
    class="c3"
  >
    React
    <span
      data-testid="subheading"
    >
       
      1 year
       
    </span>
  </div>
  <div
    class="c4 c5"
    data-testid="remove"
  />
</div>
`;

exports[`IdWrapper should be background default #80878B 1`] = `
.c2 {
  background-color: #80878B;
  color: #fff;
  padding: 5px;
}

.c3 {
  padding: 10px;
  padding-right: 80px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 span {
  color: #ccc;
}

.c5 {
  cursor: pointer;
  display: none;
}

.c0 {
  background-color: #fff;
  border: 2px solid #ccc;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 10px;
  padding: 0;
  position: relative;
}

.c0:hover .c1 {
  background-color: #1b998e;
}

.c0:hover .c4 {
  display: block;
}

@media only screen and (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    margin: 0;
  }

  .c0 .c4 {
    display: block;
  }
}

<div
  class="c0"
  data-testid="wrapper"
>
  <div
    class="c1 c2"
    data-testid="id"
  >
    1
  </div>
  <div
    class="c3"
  >
    React
    <span
      data-testid="subheading"
    >
       
      1 year
       
    </span>
  </div>
  <div
    class="c4 c5"
    data-testid="remove"
  />
</div>
`;

exports[`IdWrapper should have different background color if passes as prop 1`] = `
.c2 {
  background-color: lightgreen;
  color: #fff;
  padding: 5px;
}

.c3 {
  padding: 10px;
  padding-right: 80px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 span {
  color: #ccc;
}

.c5 {
  cursor: pointer;
  display: none;
}

.c0 {
  background-color: #fff;
  border: 2px solid #ccc;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 10px;
  padding: 0;
  position: relative;
}

.c0:hover .c1 {
  background-color: #1b998e;
}

.c0:hover .c4 {
  display: block;
}

@media only screen and (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    margin: 0;
  }

  .c0 .c4 {
    display: block;
  }
}

<div
  class="c0"
  data-testid="wrapper"
>
  <div
    class="c1 c2"
    data-testid="id"
  >
    1
  </div>
  <div
    class="c3"
  >
    React
    <span
      data-testid="subheading"
    >
       
      1 year
       
    </span>
  </div>
  <div
    class="c4 c5"
    data-testid="remove"
  />
</div>
`;
